src/Graph.cpp

```diff
@@ -148,23 +177,94 @@ std::vector<Graph::TaxiTripDetails> Graph::readCSV(std::string filename)
     // Close file
     myFile.close();
     return result;
-};
+}
 
-std::vector<Node *> *Graph::getNodes()
-{
-    return &nodes_;
+
+
+
+
+
+std::vector<std::pair<Node*, double>> Graph::findPathLengths(Node* start, int metric) {
+    std::vector<std::pair<Node*, double>> distances = 
+        std::vector<std::pair<Node*, double>>(nodes_.size());
+
+    distances.push_back(std::pair<Node*, double>(start, 0));
+    for (unsigned int i = 1; i < nodes_.size(); i++) {
+        distances.push_back(std::pair<Node*, double>(nodes_[i], DBL_MAX));
+    }
+
+    std::queue<Node*> queue;
+    queue.push(start);
+    while (!queue.empty()) {
+        Node* front = queue.front();
+        queue.pop();
+
+        for (auto neighbor : *(front->getNeighbors())) {
+            Edge* edge = neighbor.second;
+            Node* destination = neighbor.first;
+
+            int index = get_idx_path(distances, front);
+            
+            int val_to_add = 0;
+            if (metric == 0) {
+                val_to_add = edge->getFare();
+            }
+            else if (metric == 1) {
+                val_to_add = edge->getMiles();
+            }
+            else if (metric == 2) {
+                val_to_add = edge->getTime();
+            }
+            else {
+                distances[0].second = -1.0; // metric has to be 0-2.
+                return distances;
+            }
+
+
+            auto new_dist = val_to_add + get_seconds_path(distances)[index];
+            
+            index = get_idx_path(distances, destination);
+            if (new_dist < get_seconds_path(distances)[index]) {
+                distances[index].second = new_dist;
+                queue.push(destination);
+            }
+        }
+    }
+    return distances;
 }
 
-Node *Graph::getNode(std::string zone)
-{
-    for (unsigned i = 0; i < nodes_.size(); ++i)
-    {
-        if (nodes_.at(i)->getZone() == zone)
-        {
-            return nodes_.at(i);
+double Graph::findShortestPath(Node* start, Node* end, int metric) {
+    std::vector<std::pair<Node*, double>> distances = findPathLengths(start, metric);
+
+    int index = get_idx_path(distances, end);
+    return distances[index].second;
+}
+
+
+int Graph::get_idx_path(std::vector<std::pair<Node*, double>> distances, Node* destination) {
```

Comment:
this is legit getting picky, but please be consistent with the function names.  should be camel case